src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/VersionConverter.cs

```diff
@@ -1,25 +1,76 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Buffers;
+using System.Diagnostics;
+
 namespace System.Text.Json.Serialization.Converters
 {
     internal sealed class VersionConverter : JsonConverter<Version>
     {
         public override Version Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
+            if (reader._stringHasEscaping)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+#if BUILDING_INBOX_LIBRARY
+
+            ReadOnlySpan<byte> source = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;
+
+            int maxCharCount = JsonReaderHelper.s_utf8Encoding.GetMaxCharCount(source.Length);
+            char[]? pooledChars = null;
+            Span<char> charBuffer = maxCharCount * sizeof(char) <= JsonConstants.StackallocThreshold
+                ? stackalloc char[maxCharCount]
```

Comment:
I believe the guidance is to just use `JsonConstants.StackallocThreshold` for the size as the jitter can optimize when calling the method.

cc @GrabYourPitchforks

Reply:
In that case `Encoding.GetChars()` can write not complete `byte`s source to `char` buffer. Indeed, max length of string representation of valid `Version` object is 43 chars aka 86 bytes which is always lower than current `JsonConstants.StackallocThreshold` value.   
But if we want to get string representation only of valid `Version`s be successfully converted to chars, shouldn't we then just use max length of string of valid `Version` object * sizeof(char), and so we will stackalloc even less than `JsonConstants.StackallocThreshold`

Reply:
If you know that the maximum length is _x_ bytes or chars for any specific data type, it's fine to hardcode that value here. It'll result in somewhat more efficient stack usage than relying on the fallback const.

Reply:
We only know that maximum length of valid `Version` coming from CLR is 43 chars, but incoming data can be of any length, and potentially can cause exception when getting chars from encoding. To remain consistent this exception need to be catched and rethrown as `JsonException`.

```cs
string source = $"{int.MaxValue}.{int.MaxValue}.{int.MaxValue}.{int.MaxValue}";
var s_utf8Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
byte[] ut8Bytes = s_utf8Encoding.GetBytes(source);
Span<char> chars = stackalloc char[source.Length - 1]; // Buffer too small to contain all incoming data
s_utf8Encoding.GetChars(ut8Bytes, chars);
//The line above fails with unhandled exception
//Unhandled exception. System.ArgumentException:
//The output char buffer is too small to contain the decoded characters,
//encoding 'Unicode (UTF-8)' fallback 'System.Text.DecoderExceptionFallback'. (Parameter 'chars')
```